### PR TITLE
Allow different category labels in parquet partitions

### LINKED ISSUE #1836: Different categories in different partitions
Categoricals are currently difficult in dask.dataframe because all partitions must have the same set of categories.  This synchronization can be very expensive, often requiring full table scans.  Operations like `.astype('category')` or `read_csv(..., dtypes={'x': 'category'})` don't work as expected.

Recently pandas started to support union-ing categoricals on concatenation.  This ability might allow dask.dataframe to instead maintain categories per-partition, which removes the burden of synchronization.  We would only combine categoricals when different partitions need to interact.  This would help us in data loading (csv, parquet) and in aggregations (groupbys, etc.) but not for operations like `pivot_table` which expect to know the categories ahead of time.

This might be as easy as adding the right call to `union_categoricals` in `dask.dataframe.core._concat`, but there are probably other nuances.

What is the best way to do this?

cc @TomAugspurger @martindurant @sinhrks

### PR DESCRIPTION
**Pull request by martindurant**
Define the labels in the metadata using UNDEFINED_CATEGORIES.

Fixes #1836

### CODE REVIEW

File: dask/dataframe/io/tests/test_parquet.py [hunk 1]
```diff
@@ -192,3 +192,18 @@ def test_roundtrip(df, write_kwargs, read_kwargs):
...
+    ddf = dd.from_pandas(df, npartitions=2)
+    ddf['y'] = ddf.y.astype('category')
+    ddf.to_parquet(fn)
+    ddf2 = dd.read_parquet(fn, categories=['y'])
+    with pytest.raises(NotImplementedError):
+        ddf2.y.cat.categories
+    assert set(ddf2.y.compute().cat.categories) == {'a', 'b', 'c'}
+    cats_set = ddf2.map_partitions(lambda x: x.y.cat.categories).compute()
+    assert cats_set.tolist() == ['a', 'c', 'a', 'b']
+    assert ddf.compute().y.tolist() == ddf2.compute().y.tolist()
```
Comment (mrocklin) [inline dask/dataframe/io/tests/test_parquet.py]: Whenever I see `assert a == b` tests using dask.dataframes or dask.arrays I ask "why aren't we using `assert_eq` here?"  Do we need to change `assert_eq` to be aware of known and unknown categories cases?  Would the following fail?

    assert_eq(ddf, ddf2)

Comment (martindurant) [inline dask/dataframe/io/tests/test_parquet.py]: Yes, it fails - fixed it.

File: dask/dataframe/io/parquet.py [hunk 2]
```diff
@@ -121,7 +117,8 @@ def read_parquet(path, columns=None, filters=None, categories=None, index=None,
                         columns=[c for c in pf.columns if c in dtypes])
 
     for cat in categories:
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: Looks like you still need to handle the case if `categories` is `None` here. Also, what happens if the data is written as categories, but the user doesn't pass in any categories to `read_parquet`? Or if they pass in categories for a column that isn't?

File: dask/dataframe/io/parquet.py [hunk 3]
```diff
@@ -121,7 +117,8 @@ def read_parquet(path, columns=None, filters=None, categories=None, index=None,
...
-        meta[cat] = pd.Series(pd.Categorical([], categories=cats[cat]))
+        meta[cat] = pd.Series(pd.Categorical([],
+                              categories=[UNKNOWN_CATEGORIES]))
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: This looks fine to me.

File: dask/dataframe/io/parquet.py (hunk 2 above)
Comment (martindurant) [inline dask/dataframe/io/parquet.py]: If they don't include the categories, the values get de-referenced upon load.
If they pass in a column name to categories which doesn't exist, it is ignored.
I'll fix None.

Comment (jcrist) [inline dask/dataframe/io/parquet.py]: Not one that doesn't exist, I mean one that isn't a categorical. It's important to ensure that the metadata created matches exactly - no extra/missing columns, and dtypes are all correct.

Comment (martindurant) [inline dask/dataframe/io/parquet.py]: It is not possible to know whether the data is really stored as categorical ("dictionary encoding") before reading it. If the user specifies a column as categorical but any part of it is not, there will be an error on reading.

Comment (jcrist) [inline dask/dataframe/io/parquet.py]: Ah, ok. I assume the same for if there are extra columns specified?

For my own understanding, how expensive would it be to infer the encoding/columns/dtypes from the parquet file? I assume somewhere in the file there's an indication that a column is dictionary encoded?

Comment (martindurant) [inline dask/dataframe/io/parquet.py]: No extra columns given to categories are just ignored.
The encoding of any given column can vary vary by *page*, which is a sub-unit of a row-groups, so a scan of all the data would be required to say for sure that everything is dictionary encoded. For files written by fastparquet, there is only ever one page per row group, but still.

Comment (martindurant) [inline dask/dataframe/io/parquet.py]: (to answer in full: the dtype of the values are always know beforehand in the global schema, i.e., pure metadata, only the specific encoding of how that is stored - dictionary or not - is undefined at the global level)

Comment (jcrist) [inline dask/dataframe/io/parquet.py]: > No, extra columns given to categories are just ignored.

As is, extra columns are ignored by the reader, but are still put in the metadata. This causes problems, as the metadata no longer matches the dtypes/columns of each partition. Example:

```python
In [15]: df = pd.DataFrame({'a': range(5), 'b': range(5)})

In [16]: dd.from_pandas(df, npartitions=2).to_parquet('test.parq')

In [17]: ddf = dd.read_parquet('test.parq', categories=['c'])

In [18]: ddf.dtypes    # Dask thinks there's a column `c`
Out[18]:
a       int64
b       int64
c    category
dtype: object

In [19]: ddf.compute()    # Not actually a column `c`
Out[19]:
       a  b
index
0      0  0
1      1  1
2      2  2
3      3  3
4      4  4
```

Comment (martindurant) [inline dask/dataframe/io/parquet.py]: Right, got you. They should not appear, but I'm not convinced it should be an exception.

Comment (jcrist) [inline dask/dataframe/io/parquet.py]: One argument to error is that other pandas methods error on extra columns:

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'a': range(5), 'b': ['a', 'b', 'c', 'd', 'e']})

In [3]: df.to_csv('test.csv', index=False)

In [4]: pd.read_csv('test.csv', usecols=['a', 'b', 'c'])
...
ValueError: Usecols do not match names.
```

**Comment by mrocklin** [comment]
What is the status on this?

**Comment by martindurant** [comment]
Things are done, except that there is no exception if categories= in read includes a column that doesn't actually exist - it is silently ignored, as is the case for extra entries in columns, has_nulls, object_encoding, fixed_text, and the same options in fastparquet.

**Comment by jcrist** [comment]
Hmmm. I still vote that passing in extra columns is an error at graph build time (as is done elsewhere in pandas). I'm wary of silently ignoring bad inputs, as this can lead to hard to find bugs in my experience. It sounds like `fastparquet` does this though, and @martindurant is for matching this behavior. We may need a tie-breaker vote here.

**Comment by mrocklin** [comment]
Looking over the comments it sounds like we shouldn't err if a provided column name is definitely a column, but isn't stored with dictionary encoding.

There appears to be disagreement about what to do when a column name is provided that is not present at all within the dataset.  Generally I'm personally in favor of erring loudly in these cases.  My anecdotal experience is that users generally appreciate being told when they have made a mistake.

However, it also seems like this isn't standard within fastparquet.  I agree that there is value to being consistent with the choices made in that library.  So I'll extend the question over there: should fastparquet err loudly when given columns that don't exist?  I would think yes, but I have less experience with the use of that library than @martindurant

**Comment by martindurant** [comment]
We could do that in fastparquet, but it would have to be a validate_options method, since the specifications on columns and what to do with them happen in a number of places.

**Comment by martindurant** [comment]
I added the check on the fastparquet side, so we should call `fastparquet.util.check_column_names` here before attempting the read. Please check that the function there (not yet merged) does the right thing.

**Comment by martindurant** [comment]
@mrocklin , @jcrist : this should be ready now.

File: dask/dataframe/io/parquet.py [hunk 4]
```diff
@@ -120,8 +118,11 @@ def read_parquet(path, columns=None, filters=None, categories=None, index=None,
                         for (c, d) in dtypes.items()},
                         columns=[c for c in pf.columns if c in dtypes])
 
-    for cat in categories:
-        meta[cat] = pd.Series(pd.Categorical([], categories=cats[cat]))
+    for cat in categories or []:
+        if cat not in meta:
+            continue
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: This check should be no longer needed, correct?

File: dask/dataframe/io/parquet.py [hunk 5]
```diff
@@ -71,17 +73,13 @@ def read_parquet(path, columns=None, filters=None, categories=None, index=None,
...
 
+    check_column_names(pf.columns, categories)
     name = 'read-parquet-' + tokenize(pf, columns, categories)
 
     rgs = [rg for rg in pf.row_groups if
            not(fastparquet.api.filter_out_stats(rg, filters, pf.helper)) and
            not(fastparquet.api.filter_out_cats(rg, filters))]
 
-    # get category values from first row-group
-    categories = categories or []
```
Comment (jcrist) [inline dask/dataframe/io/parquet.py]: Perhaps still do this up here, as it's done a few times below?

**Comment by mrocklin** [comment]
Anything left to do here?

**Comment by martindurant** [comment]
I thought this was done.

**Comment by jcrist** [comment]
LGTM. Thanks @martindurant! Merging.

**Pull request merged**